### ISSUE TITLE
Supply gas hints to transactions in deployment script

### ISSUE THREAD

**Issue by cliffhall** [bosonprotocol/contracts#255]
## Why are we considering fiddling with the gas during deployment?
Ropsten deployment failed last week and it was gas related. The automatic gas fee hints that ethers.js was providing did not work. This _might_ be because Ropsten often has erratic gas prices, we're not certain.

However if we provide the gas hints at deployment time, we can check the prevailing price and add enough wiggle room that they'll likely all go through.

**Comment by cliffhall** [comment]
TL;dr - For post-London (ETH2) style transactions, [Alchemy recommends](https://docs.alchemy.com/alchemy/guides/eip-1559/maxpriorityfeepergas-vs-maxfeepergas) setting just the `maxPriorityFeePerGas` in order to ensure the transaction gets into the block.

There are 3 knobs now. `baseFeePerGas`, `maxPriorityFeePerGas`, and `maxFeePerGas`.

Base fee is set by the system, and all transactions in the block will pay it. You can specify a tip or you can specify a ceiling which includes the base fee and tip. By setting the tip only, the other two are automatically set right and the miner is ensured the tip (_bribe_) amount.

<img width="780" alt="Screen Shot 2021-11-01 at 3 41 51 PM" src="https://user-images.githubusercontent.com/871933/139731470-04e69208-bd19-40d4-a70e-30b9a6d4ce26.png">

**Comment by cliffhall** [comment]
## Pricing Strategy
We will need to get an estimate of what to use for `maxPriorityFeePerGas`.

I have used etherchain.org's [Gas Price Oracle](https://www.etherchain.org/tools/gasPriceOracle) for over a year and it's stable. It has an API and can be automated against reliably, with a sane fallback strategy (I've done this on several projects).

They show the current base fee and a suggested max fee or tip. The max fee strategy seems to be _double the base, add max tip, and add a pinch more._ As discussed in previous comment, we won't be supplying such a value. Instead we'll just supply the tip. 

We'll use whatever the gwei value for "Fastest" is. Not because we're concerned about speed, but because it gives us the most ability to handle 'slippage' if the tip we set at the beginning of the deployment script becomes less viable before completion. Note the considerable difference between the "Fastest" tip and the "Fast" tip. That's a lot. If both numbers doubled during deployment, the max tip we supplied will still be more than enough.

<img width="1392" alt="Screen Shot 2021-11-01 at 4 46 48 PM" src="https://user-images.githubusercontent.com/871933/139739875-c41a81d4-9cc7-4bf9-ac9a-f16c8501ffc6.png">

**Comment by cliffhall** [comment]
## How will we specify the tip during deployment?
We need to make sure every transaction gets the news that we want to set `maxPriorityFeePerGas`.

The easy way would be to configure that in the `hardhat.config` file. Unfortunately, that's[ not currently settable](https://github.com/nomiclabs/hardhat/issues/1866). 

Consequently, we need to add the options block to all contract `deploy()` and `contractMethod()` transactions.

This is verbose, but the end effect:
```javascript
this.tokenRegistry = await TokenRegistry.deploy({maxPriorityFeePerGas: this.maxTip});
await this.tokenRegistry.setTokenLimit(this.boson_token, this.TOKEN_LIMIT, {maxPriorityFeePerGas: this.maxTip});
```

So, in the `DeploymentExecutor` constructor, we'll set the options as an instance property.

```javascript
    this.maxTip = ethers.utils.parseUnits(
      process.env.MAX_TIP
        ? String(process.env.MAX_TIP)
        : "1"
      ,"gwei");

    this.txOptions = {maxPriorityFeePerGas: this.maxTip};
```

Then we can make those calls a bit more legible.
```javascript
this.tokenRegistry = await TokenRegistry.deploy(this.txOptions);
await this.tokenRegistry.setTokenLimit(this.boson_token, this.TOKEN_LIMIT, this.txOptions);
```

Finally, we'll manually check the [Gas Price Oracle](https://www.etherchain.org/tools/gasPriceOracle) before deployment and set the "Fastest" value in our `.env` file. 

```javascript
// Max tip in gwei - use "Fastest" setting from: https://www.etherchain.org/api/gasPriceOracle
MAX_TIP = 13.1
```
